Application: git source

vgramer opened the pull request:
**What does this PR do / Why do we need it**:

Add code to download the sources from a git repository

**Does this PR close any issues?**:<!-- optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged -->
Fixes #8379 

**Special notes for your reviewer**:

**Documentation**:
<!-- Add links to the related documentation changes related to this pull request. E.g. the link to the kubermatic/docs pull request. -->

**Does this PR introduce a user-facing change?**:
<!-- Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

vgramer committed:
application: add gitSource

GitSource download the application source from a git repository.

Signed-off-by: Vincent Gramer <vincent@kubermatic.com>

vgramer committed:
application: add tests for gitSource

Signed-off-by: Vincent Gramer <vincent@kubermatic.com>

ahmedwaleedmalik commented on pkg/applications/providers/types.go:
```diff
@@ -0,0 +1,48 @@
+func NewSourceProvider(ctx context.Context, client ctrlruntimeclient.Client, appSource *appskubermaticv1.ApplicationSource, secretNamespace string) (SourceProvider, error) {
+	switch {
+	case appSource.Helm != nil:
+		return nil, fmt.Errorf("helmSource is not inmpleted yet")
```
`s/inmpleted/implemented`

ahmedwaleedmalik commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+	return err
+}
+
+// checkoutFromTag clone the repository and checkout the desired Tag. A swallow clone is performed.
```
`s/swallow/shallow`

vgramer replied:
i also updated the one in tests too :)

ahmedwaleedmalik commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+	return workTree.Checkout(&git.CheckoutOptions{Hash: commit.Hash})
+}
+
+// checkoutFromBranch clone the repository and checkout the desired branch. A swallow clone is performed.
```
`s/swallow/shallow`

ahmedwaleedmalik commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+	}
+
+	checkoutStrategy := g.getCheckoutStrategy()
+	if err := checkoutStrategy(g.Ctx, destination, g.Source, auth); err != nil {
```
nit

`checkoutStrategy(g.Ctx, destination, g.Source, auth)` looks a bit misleading. Since we are actually performing a `git checkout` here. Maybe we can do something like `checkout := g.getCheckoutStrategy()`

P.s. My concern here is just code coherence, nothing logical or syntactical.

SimonTheLeg commented on cmd/user-cluster-controller-manager/main.go:
```diff
@@ -382,7 +382,7 @@ func main() {
 	}
 
-	if err := applicationinstallationcontroller.Add(rootCtx, log, seedMgr, mgr, isPausedChecker, &applications.ApplicationManager{}); err != nil {
+	if err := applicationinstallationcontroller.Add(rootCtx, log, seedMgr, mgr, isPausedChecker, &applications.ApplicationManager{SecretNamespace: runOp.namespace}); err != nil {
```
q: So it was right to use the runOp.namespace, because it just gets propagated from master through seed to user, right?

vgramer replied:
As you know, application credentials are replicated from master to seed clusters into "kubermatic" namespace by the `application-secret-synchronizer`. 

The user cluster does not have the permission to read secrets into the "kubermatic" namespace.  I could add RBAC to provide access to user-cluster-controller-manager to the kubermatic namespace. Nevertheless, it feels wrong as this namespace has many sensible secrets (especially on master/seed setup). 

After discussing with @sig-cluster-management, we agree the better solution is to have a new controller in the seed-controller-manager that replicates application secrets into all cluster namespaces (e.g cluster-znx6n7jbdz). (it will be implemented in another PR...)

`runOp.Namespace`is the namespace where the user-cluster-controller is running (ie cluster-<cluster_id>).  Currently for testing, I manually copy application credentials into the cluster namespace....

SimonTheLeg replied:
ah okay understood.

SimonTheLeg commented on go.mod:
```diff
@@ -51,10 +52,12 @@ require (
 	github.com/prometheus/client_golang v1.12.1
 	github.com/robfig/cron/v3 v3.0.1
 	github.com/sirupsen/logrus v1.8.1
+	github.com/sosedoff/gitkit v0.3.0
```
is there a reason we are adding gitkit? I don't see it being imported anywhere.

SimonTheLeg replied:
Same goes for freeport

vgramer replied:
it used for testing cf https://github.com/kubermatic/kubermatic/pull/9977#discussion_r892469553

SimonTheLeg replied:
ah now I get it. The test file is collapsed by default. Okay makes sense

SimonTheLeg commented on pkg/applications/installer.go:
```diff
@@ -42,14 +44,35 @@ type ApplicationInstaller interface {
+	if err != nil {
+		return fmt.Errorf("failed to create temp directory where application source will be downloaded: %w", err)
+	}
+	defer os.RemoveAll(downloadDest)
```
since this can throw an error, we should at least log the error. So we can see if there might be a bug filling up the volume of the pod

SimonTheLeg replied:
Since this can throw an error, I think we should at least log the error. This could help us in production if something unexpected happens

SimonTheLeg commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+type checkoutFunc func(ctx context.Context, destination string, gitSource *appskubermaticv1.GitSource, auth transport.AuthMethod) error
+
+// checkoutFromCommit clone the repository and checkout the desired commit.
+// If a branch is defined then a shallow clone of the branch is performed. The commit must belongs to this branch.
```
```suggestion
// If a branch is defined then a shallow clone of the branch is performed. The commit must belong to this branch.
```

vgramer replied:
done :)

SimonTheLeg commented on pkg/applications/installer.go:
```diff
@@ -42,14 +44,35 @@ type ApplicationInstaller interface {
+		return fmt.Errorf("failed to download application source: %w", err)
+	}
+
+	log.Debugw("application download", "location", appSourcePath)
```
```suggestion
	log.Debugw("application successfully downloaded", "location", appSourcePath)
```

SimonTheLeg commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+	"github.com/go-git/go-git/v5/plumbing"
+	"github.com/go-git/go-git/v5/plumbing/transport"
+	"github.com/go-git/go-git/v5/plumbing/transport/http"
+	"github.com/go-git/go-git/v5/plumbing/transport/ssh"
```
as these have quite generic names, I think it might make sense to import them prefaced with "gogit":

```
gogittransport "github.com/go-git/go-git/v5/plumbing/transport"
```

Especially I think for funcs like authFromCredentials to make it easier for the consumer

vgramer replied:
done :)

SimonTheLeg commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+}
+
+// authFromCredentials returns a new transport.AuthMethod according to credentials defined in the GitSource.
+// If no crednetials are defined in the GitSource then nil is returned.
```
```suggestion
// If no credentials are defined in the GitSource then nil is returned.
```

SimonTheLeg commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+			if err != nil {
+				return nil, fmt.Errorf("failed to parse private ssh key: %w", err)
+			}
+			authssh.HostKeyCallback = gossh.InsecureIgnoreHostKey()
```
I guess for now it is okay to allow any host. Maybe we should add a small section to the explanation of the func, that for now all hosts are accepted by default

SimonTheLeg commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+			SingleBranch:  true,
+			ReferenceName: plumbing.NewBranchReferenceName(gitSource.Ref.Branch),
+			Tags:          git.NoTags,
+			NoCheckout:    true,
```
Now that I see this again, I am wondering if this is not too complicated for the user. Essentially this creates a lot of possible states:
* branch only → super fast, depth 1 shallow clone
* branch+commit → medium fast, depth max, but only for one branch shallow clone
* commit only → full clone, either medium fast (if not too many other branches) or slow
* tag only → super fast, depth 1 shallow clone

In the end, this is due to git's design, which does not allow you to shallow clone a commit. And we know this, but the end-user doesn't. Therefore I think, we should at least re-word the user facing section here https://github.com/vgramer/kubermatic/blob/5c835b66f7315b1cbbcd6257a5e32f91a15a5c96/pkg/apis/apps.kubermatic/v1/application_definition.go#L84
to include the information.

Maybe we could do something like:
"For large repositories we recommend to either use Tag, Branch+Commit, or Branch. This allows a shallow clone, which dramatically speeds up performance"

SimonTheLeg commented on pkg/applications/providers/source/git_test.go:
```diff
@@ -0,0 +1,701 @@
+//     └── file5
+func createGitRepository(t *testing.T, temp string) repoInfo {
+	repoDir := temp + "/repo1"
+	err := os.Mkdir(repoDir, 0700)
```
mega nitpick, but I think we can completely create this in memory. afaik both go-git as well as go stdlib now support io/fs, which has an implementation for in-memory systems.

But again, just a nitpick :)

vgramer replied:
In fact, it's a bit more complicated. I use go-git to create a local repository and easily keep track of references needed for tests (tag, master branch ...) then I use `gitkit` to start a local git server that serves this repository over HTTP (with and without auth) and SSH.  So this repository can not be in memory.

notes:
* freeport is used to allocate a freeport when git server serves the repository over ssh.
* gitkit only need a bare git repository. I guess I could use the go-git plumbing command to create a bare repository but creating a plane one was easier....

SimonTheLeg reviewed (changes requested):
only some smaller changes. Overall this is a great PR. Everything is documented and very good to understand. And testing is 👌

vgramer committed:
apply review comments

Signed-off-by: Vincent Gramer <vincent@kubermatic.com>

embik commented on pkg/applications/installer.go:
```diff
@@ -42,14 +44,39 @@ type ApplicationInstaller interface {
+		return fmt.Errorf("failed to initialize source provider: %w", err)
+	}
+
+	downloadDest, err := os.MkdirTemp("", applicationInstallation.Namespace+"-"+applicationInstallation.Namespace)
```
This can potentially create HUGE tmp directories depending on the git source. We need to put in safeguards, like using an empty dir in the user-cluster-controller-manager Deployment, and limiting the amount of disk space this can eat. Otherwise you can fully consume the underlying node's disk.

embik replied:
(I know there's a shallow clone happening but IMHO we should still make sure this gets its own temporary volume with storage limits on it)

vgramer replied:
you're totally right, i forgot we don't have default limits :s  working on it.

vgramer replied:
@embik 
I have added an empty dir with a default size is 300Mi.

To choose this default value, I take a look at the size of 3 repositories:
* https://github.com/bitnami/charts: set of ~100 charts provided by bitnami
* https://github.com/prometheus-community/helm-charts: contains the different charts for deploying prometheus stack
* https://github.com/kubermatic/kubermatic

I choose these repositories because the first 2 serve only charts and the last one serves charts and code (with history).

|   repository         | clone all repo   | clone a tag  |
|-----------------|----------------|-------------|
| bitanmi               | 199M                 | 38M             |
| prometheus       | 87M                   | 6,7M             |
| kubermatic.       | 191M                  | 50M             |

So a 300Mi cache should be more than enough to download sources from a repo and handle dependencies.

_note: the max reconciliation of the application controller is 1._

embik replied:
Great, thanks for adding that!

vgramer committed:
application: donwload sources in an emptyDir

the size of the empty dir can be set through Cluster.spec.applicationSettings.CacheSize. The default size is 300Mi.

To choose this default value, we looks a the size of 3 repositories:
* https://github.com/bitnami/charts: set of ~100 charts provided by bitnami
* https://github.com/prometheus-community/helm-charts: contains the different charts for deploying prometheus stack
* https://github.com/kubermatic/kubermatic

These repositories has been choosen because, the first 2 serve only charts and the last one serve charts and code (with history).

repository              size with clone commit          size with clone tag
bitnami                 199m                            38m
prometheus              87m                             6,7m
kubermatic              191m                            50m

So with a 300Mi cache should be more than enough to donwload source from a big repo and handle dependencies

Signed-off-by: Vincent Gramer <vincent@kubermatic.com>

vgramer committed:
add go-git as dependency(used to interact with git repository) and add  gitkit, freeport(used for testing)

Signed-off-by: Vincent Gramer <vincent@kubermatic.com>

vgramer committed:
update fixtures

Signed-off-by: Vincent Gramer <vincent@kubermatic.com>

ahmedwaleedmalik commented on hack/run-user-cluster-controller-manager.sh:
```diff
@@ -108,4 +108,5 @@ set -x
   -seed-kubeconfig=${SEED_KUBECONFIG} \
   -owner-email=${OWNER_EMAIL} \
   -dns-cluster-ip=10.240.16.10 \
+  -application-cache="${TMPDIR}" \
```
Wouldn't a dedicated temp directory for this make more sense for this? For reference https://github.com/kubermatic/kubermatic/blob/master/hack/lib.sh#L198

ahmedwaleedmalik commented on pkg/apis/apps.kubermatic/v1/application_definition.go:
```diff
@@ -106,7 +106,9 @@ type GitSource struct {
-	// Git reference to checkout
+	// Git reference to checkout.
+	//
+	// For large repositories, we recommend to either use Tag, Branch or Branch+Commit. This allows a shallow clone, which dramatically speeds up performance"
```
```suggestion
	// For large repositories, we recommend to either use Tag, Branch or Branch+Commit. This allows a shallow clone, which dramatically speeds up performance
```

ahmedwaleedmalik commented on pkg/applications/installer.go:
```diff
@@ -42,14 +44,42 @@ type ApplicationInstaller interface {
 
 // ApplicationManager handles the installation / uninstallation of an Application on the user-cluster.
 type ApplicationManager struct {
+	// ApplicationCache is the path to the directory used for chaching applications. (i.e. locaction where application's source will be downloaded, Helm repository cache ...)
```
`s/chaching/caching`
`s/locaction/location`

ahmedwaleedmalik commented on pkg/applications/installer.go:
```diff
@@ -42,14 +44,42 @@ type ApplicationInstaller interface {
+		return fmt.Errorf("failed to initialize source provider: %w", err)
+	}
+
+	downloadDest, err := os.MkdirTemp(a.ApplicationCache, applicationInstallation.Namespace+"-"+applicationInstallation.Namespace)
```
```suggestion
	downloadDest, err := os.MkdirTemp(a.ApplicationCache, applicationInstallation.Namespace+"-"+applicationInstallation.Name)
```

vgramer replied:
oh shit :s  nice catch!

ahmedwaleedmalik commented on pkg/applications/providers/source/git.go:
```diff
@@ -0,0 +1,217 @@
+		if err != nil {
+			return err
+		}
+	}
```
Just a nit; you can skip this if you want to.

I think for code coherence, we can resort to something like this:

```go
	cloneOpts := git.CloneOptions{
		URL:          gitSource.Remote,
		Auth:         auth,
		Progress:     nil,
		Tags:         git.NoTags,
		NoCheckout:   true,
	}

	if len(gitSource.Ref.Branch) > 0 {
		// We need to checkout a commit form specific branch
		cloneOpts.SingleBranch = true
		cloneOpts.ReferenceName = plumbing.NewBranchReferenceName(gitSource.Ref.Branch),
	} else {
		// We need to checkout a commit from head
		cloneOpts.SingleBranch = false
	}

	repo, err = git.PlainCloneContext(ctx, destination, false, cloncloneOpts)
	if err != nil {
		return err
	}
```

ahmedwaleedmalik replied:
Also, can't we use `SingleBranch` if branch is not specified? Because in that case `HEAD` will be used.

vgramer replied:
> I think for code coherence, we can resort to something like this:

agree with you :)

> Also, can't we use SingleBranch if branch is not specified? Because in that case HEAD will be used.

No, because if the user specifies a commit we should look for it in all history. As mentioned earlier it's not recommended to clone only with commit but it can be useful in some cases.

ahmedwaleedmalik replied:
> we should look for it in all history. 

By that, you mean all branches or just the default branch?

vgramer replied:
all branches

ahmedwaleedmalik replied:
That's not right then. If user only specifies the `commit hash` then we should assume that it's from the master/main branch.

vgramer replied:
why should we assume that?

ahmedwaleedmalik replied:
As discussed, we'll create a follow up ticket to discuss this.

vgramer committed:
apply review comments

Signed-off-by: Vincent Gramer <vincent@kubermatic.com>

ahmedwaleedmalik reviewed (approved):
/approve 🚀 

Let's get back and resolve(if required) https://github.com/kubermatic/kubermatic/pull/9977#discussion_r915762004 based on the consensus in `sig-cluster-management`, in a follow-up PR. It's better to at least get these parts merged.
